### PR TITLE
fix: draft recommendation status to show correct value

### PR DESCRIPTION
**Pull request by jayanta2018**
- Check draft recommendation: if yes return draft else normal **_original_** flow
- Edge case: check if the Recommendation is submitted by other means, i.e., GMC Connect (if GMC submission date is past and ignore it, i.e. return empty TraineeRecommendationRecordDto)

### CODE REVIEW

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceImpl.java [hunk 1]
```diff
@@ -283,26 +284,43 @@ public boolean submitRecommendation(String recommendationId, String gmcNumber,
...
+        .collect(Collectors.toList());
+    //Check draft recommendation: if yes return draft else normal original flow
+    if (draftRecommendations.size() > 0) {
+      final var draftRecommendation = draftRecommendations.get(0);
+      //Edge case: check if the Recommendation is submitted by other means, i.e., GMC Connect
+      //if GMC submission date is past and ignore it, i.e. return empty TraineeRecommendationRecordDto
+      if (draftRecommendation.getGmcSubmissionDate().isBefore(LocalDate.now())) {
+        return new TraineeRecommendationRecordDto();
+      }
+      draftRecommendation.setRecommendationStatus(RecommendationStatus.DRAFT);
```
Comment (hihilary) [inline application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceImpl.java:299]: Why do we set the recommendation status to DRAFT?
In RecommendationStatus enumeration, we have DRAFT, but do we really use it?

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceImpl.java [hunk 2]
```diff
@@ -283,26 +284,37 @@ public boolean submitRecommendation(String recommendationId, String gmcNumber,
...
-    if (!optionalDoctorsForDB.isPresent()) {
-      throw new RecommendationException(format(DOCTOR_NOT_FOUND_MESSAGE, gmcId));
-    }
+    final var recommendations = recommendationRepository.findByGmcNumber(gmcId);
+    final var draftRecommendations = recommendations.stream().filter(
+        recommendation -> recommendation.getRecommendationStatus().name().equals("READY_TO_REVIEW"))
+        .collect(Collectors.toList());
+    //Check draft recommendation: if yes return draft else normal original flow
+    if (draftRecommendations.size() > 0) {
+      final var draftRecommendation = draftRecommendations.get(0);
```
Comment (hihilary) [inline application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceImpl.java:293]: Do we assume to have only 1 draft recommendation for every doctor?
Could we analyse the Prod data to check if this assumption is true, please?

File: application/src/test/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceTest.java [hunk 3]
```diff
@@ -189,6 +192,9 @@ public void setup() {
     firstName = faker.name().firstName();
     lastName = faker.name().lastName();
     status = NOT_STARTED;
+    draftRecommendationInitialStatus = READY_TO_REVIEW;
+    draftRecommendationStatus = DRAFT;
```
Comment (hihilary) [inline application/src/test/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceTest.java:196]: This status is never used.

File: application/src/test/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceTest.java [hunk 4]
```diff
@@ -189,6 +192,9 @@ public void setup() {
...
+    gmcSubmissionDate = LocalDate.now().minusDays(10);
```
Comment (hihilary) [inline application/src/test/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceTest.java:197]: This date is not used in the test.

File: application/src/test/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceTest.java [hunk 5]
```diff
@@ -788,6 +794,36 @@ void shouldReturnLatestRecommendationsList() {
     assertThat(traineeRecommendationRecordDto.getGmcNumber(), is(nullValue()));
   }
 
+  @Test
+  void shouldCheckDraftRecommendations() {
+    final var gmcNumber2 = faker.number().digits(7);
```
Comment (hihilary) [inline application/src/test/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceTest.java:799]: Can't see any assertions for the data `gmcNumber2`, so I think this is not necessary.

File: application/src/test/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceTest.java [hunk 6]
```diff
@@ -788,6 +794,36 @@ void shouldReturnLatestRecommendationsList() {
...
+    final var gmcNumberX = faker.number().digits(7);
+    final var recommendation = buildRecommendation(gmcNumber1, recommendationId,
+        draftRecommendationInitialStatus,
+        UNDER_REVIEW);
+    final var draftRecommendations = List.of(recommendation);
+
+    when(doctorsForDBRepository.findById(any())).thenReturn(Optional.of(doctorsForDB1));
+    when(recommendationRepository.findByGmcNumber(gmcNumber1))
+        .thenReturn(draftRecommendations);
```
Comment (hihilary) [inline application/src/test/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceTest.java:808]: There're no stubbing for data `gmcNumber2` & `gmcNumberX`

**Review by hihilary — CHANGES_REQUESTED** [review]

File: application/src/test/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceTest.java (hunk 3 above)
Comment (jayanta2018) [reply application/src/test/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceTest.java:196]: Removed

File: application/src/test/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceTest.java (hunk 4 above)
Comment (jayanta2018) [reply application/src/test/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceTest.java:197]: Removed

File: application/src/test/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceTest.java (hunk 5 above)
Comment (jayanta2018) [reply application/src/test/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceTest.java:795]: Used in the assertion now

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceImpl.java (hunk 2 above)
Comment (jayanta2018) [reply application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceImpl.java:293]: Test is modified with 2 draft and 1 NOT_STARTED recommendation

Comment (hihilary) [reply application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceImpl.java:293]: If there're multiple draft recommendations, may I ask how you know which one is the latest?

Comment (hihilary) [reply application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationServiceImpl.java:292]: We've check there're only 1 doctor with multiple(2) draft recommendations. And those 2 recommendations are exactly the same. We agreed to ignore it for the moment.

**Review by hihilary — APPROVED** [review]

**Pull request merged**